File: l138.py
"""Copy List with Random Pointer"""

# Definition for singly-linked list with a random pointer.
class RandomListNode(object):
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None


class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        # """
        # if head is None:
        #     return None
        # ocurr = head
        # newhead = RandomListNode(head.label)
        # ncurr = newhead
        # looktable = {head:newhead, None:None}
        # while ocurr.next is not None:
        #     ncurr.next = RandomListNode(ocurr.next.label)
        #     ncurr = ncurr.next
        #     ocurr = ocurr.next
        #     looktable[ocurr] = ncurr
        # ocurr = head
        # ncurr = newhead
        # while ocurr is not None:
        #     ncurr.random = looktable[ocurr.random]
        #     ncurr = ncurr.next
        #     ocurr = ocurr.next
        # return newhead

        if head is None:
            return None
        curr = head
        mapdic = {None:None}
        while curr:
            mapdic[curr] = RandomListNode(curr.label)
            curr = curr.next
        curr = head
        while curr:
            mapdic[curr].next = mapdic[curr.next]
            mapdic[curr].random = mapdic[curr.random]
            curr = curr.next
        return mapdic[head]

        



        

File: OAs/Amazon/l138.java
import java.util.HashMap;

// Copy List with Random Pointer

/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution {
    public RandomListNode copyRandomList(RandomListNode head) {
        if (head == null) {
            return null;
        }
        HashMap<RandomListNode, RandomListNode> map = new HashMap<>();
        RandomListNode curr = head;
        while (curr != null) {
            RandomListNode copyedNode = new RandomListNode(curr.label);
            map.put(curr, copyedNode);
            curr = curr.next;
        }

        curr = head;

        while (curr != null) {
            RandomListNode node = map.get(curr);
            node.next = map.get(curr.next);
            node.random = map.get(curr.random);
            curr = curr.next; 
        }

        return map.get(head);
    }
}